File: interfaceTest/ams/ssros_ams_py/ssros_ams_py/spin_action_server.py
import time
import rclpy
from rclpy.action import ActionServer, CancelResponse, GoalResponse
from rclpy.node import Node
from test_interfaces.action import Spin


class SpinActionServer(Node):
    def __init__(self):
        super().__init__("spin_action_server_py")
        self._action_server = ActionServer(
            self,
            Spin,
            "spin",
            self.execute_callback,
            goal_callback=self.goal_callback,
            cancel_callback=self.cancel_callback,
        )

    def goal_callback(self, goal_request):
        self.get_logger().info(
            "Received goal request with target yaw %f" % goal_request.target_yaw
        )
        return GoalResponse.ACCEPT

    def cancel_callback(self, goal_handle):
        self.get_logger().info("Received request to cancel goal")
        return CancelResponse.ACCEPT

    def execute_callback(self, goal_handle):
        self.get_logger().info("Executing goal...")
        feedback_msg = Spin.Feedback()
        feedback_msg.angular_distance_traveled = 0.0

        for i in range(int(goal_handle.request.target_yaw)):
            if goal_handle.is_cancel_requested:
                goal_handle.canceled()
                self.get_logger().info("Goal Canceled")
                return Spin.Result()

            feedback_msg.angular_distance_traveled = float(i)
            self.get_logger().info("Publish Feedback")
            goal_handle.publish_feedback(feedback_msg)
            time.sleep(1)

        goal_handle.succeed()
        self.get_logger().info("Goal Succeeded")

        result = Spin.Result()
        result.error_code = Spin.Result.NONE
        return result


def main(args=None):
    rclpy.init(args=args)
    spin_action_server = SpinActionServer()
    rclpy.spin(spin_action_server)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: interfaceTest/ams/ssros_ams/src/spin_action_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "test_interfaces/action/spin.hpp"

#include <memory>
#include <thread>

class SpinActionServer : public rclcpp::Node
{
public:
  using Spin = test_interfaces::action::Spin;
  using GoalHandleSpin = rclcpp_action::ServerGoalHandle<Spin>;

  explicit SpinActionServer(const rclcpp::NodeOptions & options = rclcpp::NodeOptions())
  : Node("spin_action_server", options)
  {
    using namespace std::placeholders;

    this->action_server_ = rclcpp_action::create_server<Spin>(
      this,
      "spin",
      std::bind(&SpinActionServer::handle_goal, this, _1, _2),
      std::bind(&SpinActionServer::handle_cancel, this, _1),
      std::bind(&SpinActionServer::handle_accepted, this, _1));
  }

private:
  rclcpp_action::Server<Spin>::SharedPtr action_server_;

  rclcpp_action::GoalResponse handle_goal(
    const rclcpp_action::GoalUUID & uuid,
    std::shared_ptr<const Spin::Goal> goal)
  {
    RCLCPP_INFO(this->get_logger(), "Received goal request with target yaw %f", goal->target_yaw);
    (void)uuid;
    return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
  }

  rclcpp_action::CancelResponse handle_cancel(
    const std::shared_ptr<GoalHandleSpin> goal_handle)
  {
    RCLCPP_INFO(this->get_logger(), "Received request to cancel goal");
    (void)goal_handle;
    return rclcpp_action::CancelResponse::ACCEPT;
  }

  void handle_accepted(const std::shared_ptr<GoalHandleSpin> goal_handle)
  {
    using namespace std::placeholders;
    // this needs to return quickly to avoid blocking the executor, so spin up a new thread
    std::thread{std::bind(&SpinActionServer::execute, this, _1), goal_handle}.detach();
  }

  void execute(const std::shared_ptr<GoalHandleSpin> goal_handle)
  {
    RCLCPP_INFO(this->get_logger(), "Executing goal");
    rclcpp::Rate loop_rate(1);
    const auto goal = goal_handle->get_goal();
    auto feedback = std::make_shared<Spin::Feedback>();
    auto & angular_distance_traveled = feedback->angular_distance_traveled;
    angular_distance_traveled = 0;
    auto result = std::make_shared<Spin::Result>();

    for (int i = 0; (i < goal->target_yaw) && rclcpp::ok(); ++i) {
      // Check if there is a cancel request
      if (goal_handle->is_canceling()) {
        result->error_code = Spin::Result::UNKNOWN;
        goal_handle->canceled(result);
        RCLCPP_INFO(this->get_logger(), "Goal Canceled");
        return;
      }
      // Update sequence
      angular_distance_traveled = i;
      // Publish feedback
      goal_handle->publish_feedback(feedback);
      RCLCPP_INFO(this->get_logger(), "Publish Feedback");

      loop_rate.sleep();
    }

    // Check if goal is done
    if (rclcpp::ok()) {
      result->error_code = Spin::Result::NONE;
      goal_handle->succeed(result);
      RCLCPP_INFO(this->get_logger(), "Goal Succeeded");
    }
  }
};

int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);
    auto action_server = std::make_shared<SpinActionServer>();
    rclcpp::spin(action_server);
    rclcpp::shutdown();
    return 0;
}